File: dashboard/src/lib/contracts/sdk-kits/paygate-client.js
// @ts-nocheck
import crypto from "node:crypto";

export class PayGateAPIError extends Error {
	constructor(message, details = {}) {
		super(message);
		this.name = "PayGateAPIError";
		this.statusCode = details.statusCode;
		this.code = details.code;
		this.requestId = details.requestId;
	}
}

export class PayGateClient {
	constructor({ baseURL, apiToken, timeoutMs = 10000 }) {
		if (!baseURL) throw new Error("baseURL is required");
		if (!apiToken) throw new Error("apiToken is required");

		this.baseURL = baseURL.replace(/\/+$/, "");
		this.apiToken = apiToken;
		this.timeoutMs = timeoutMs;
	}

	async charge(payload, options = {}) {
		return this.request("POST", "/v1/transactions/charge", {
			body: payload,
			headers: options.idempotencyKey
				? {
						"Idempotency-Key": options.idempotencyKey,
					}
				: {},
		});
	}

	async getTransaction(orderId) {
		return this.request("GET", `/v1/transactions/${encodeURIComponent(orderId)}`);
	}

	async listAuditLogs(params = {}) {
		const query = new URLSearchParams();
		for (const [key, value] of Object.entries(params)) {
			if (value === undefined || value === null || value === "") continue;
			query.set(key, String(value));
		}

		const path = query.size > 0 ? `/v1/audit-logs?${query.toString()}` : "/v1/audit-logs";
		return this.request("GET", path);
	}

	verifyWebhook({ rawBody, timestamp, signature, webhookSecret, maxSkewSeconds = 300 }) {
		if (!rawBody || !timestamp || !signature || !webhookSecret) {
			throw new Error("rawBody, timestamp, signature, and webhookSecret are required");
		}

		const unix = Number(timestamp);
		if (!Number.isFinite(unix)) {
			throw new Error("Webhook timestamp is not valid");
		}

		if (Math.abs(Math.floor(Date.now() / 1000) - unix) > maxSkewSeconds) {
			throw new Error("Webhook timestamp is too old");
		}

		const expected =
			"sha256=" +
			crypto.createHmac("sha256", webhookSecret).update(`${timestamp}.${rawBody}`).digest("hex");

		if (!crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(expected))) {
			throw new Error("Webhook signature is not valid");
		}

		return JSON.parse(rawBody);
	}

	async request(method, path, options = {}) {
		const controller = new AbortController();
		const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
		const headers = new Headers({
			Authorization: `Bearer ${this.apiToken}`,
			Accept: "application/json",
			...options.headers,
		});

		if (options.body !== undefined) {
			headers.set("Content-Type", "application/json");
		}

		try {
			const response = await fetch(`${this.baseURL}${path}`, {
				method,
				headers,
				body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
				signal: controller.signal,
			});

			const payload = await response.json().catch(() => null);
			if (!response.ok || payload?.success === false) {
				throw new PayGateAPIError(payload?.error?.message ?? "PayGate request failed", {
					statusCode: response.status,
					code: payload?.error?.code,
					requestId: payload?.error?.request_id,
				});
			}

			return payload?.data ?? null;
		} finally {
			clearTimeout(timeout);
		}
	}
}

export async function exampleUsage() {
	const client = new PayGateClient({
		baseURL: "https://paygate.digixsolution.net",
		apiToken: process.env.PAYGATE_STORE_API_TOKEN,
	});

	const transaction = await client.charge(
		{
			order_id: "INV-2026-0001",
			amount: 150000,
			currency: "IDR",
			payment_type: "bank_transfer",
			bank: "bsi",
			customer: {
				name: "Budi",
				email: "budi@example.com",
				phone: "+628123456789",
			},
			items: [
				{
					id: "SKU-1",
					name: "Kaos PayGate",
					qty: 1,
					price: 150000,
				},
			],
		},
		{
			idempotencyKey: "idem_INV-2026-0001",
		},
	);

	console.log(transaction);
}

File: dashboard/src/lib/contracts/sdk-kits/paygate-client.rs
use anyhow::{anyhow, Result};
use hmac::{Hmac, Mac};
use reqwest::{Client as HttpClient, Method};
use serde_json::{json, Value};
use sha2::Sha256;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

type HmacSha256 = Hmac<Sha256>;

pub struct PayGateClient {
    base_url: String,
    api_token: String,
    http: HttpClient,
}

impl PayGateClient {
    pub fn new(base_url: impl Into<String>, api_token: impl Into<String>) -> Result<Self> {
        Ok(Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            api_token: api_token.into(),
            http: HttpClient::builder()
                .timeout(Duration::from_secs(10))
                .build()?,
        })
    }

    pub async fn charge(&self, payload: Value, idempotency_key: Option<&str>) -> Result<Value> {
        let mut request = self.request(Method::POST, "/v1/transactions/charge");
        if let Some(value) = idempotency_key {
            request = request.header("Idempotency-Key", value);
        }

        self.send(request.json(&payload)).await
    }

    pub async fn get_transaction(&self, order_id: &str) -> Result<Value> {
        self.send(self.request(
            Method::GET,
            &format!("/v1/transactions/{}", urlencoding::encode(order_id)),
        ))
        .await
    }

    pub async fn list_audit_logs(&self, params: &[(&str, &str)]) -> Result<Value> {
        self.send(self.request(Method::GET, "/v1/audit-logs").query(params))
            .await
    }

    pub fn verify_webhook(
        &self,
        raw_body: &str,
        timestamp: &str,
        signature: &str,
        webhook_secret: &str,
        max_skew_seconds: u64,
    ) -> Result<Value> {
        let unix: u64 = timestamp.parse().map_err(|_| anyhow!("Webhook timestamp is not valid"))?;
        let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
        let delta = now.abs_diff(unix);
        if delta > max_skew_seconds {
            return Err(anyhow!("Webhook timestamp is too old"));
        }

        let mut mac = HmacSha256::new_from_slice(webhook_secret.as_bytes())?;
        mac.update(format!("{timestamp}.{raw_body}").as_bytes());
        let expected = format!("sha256={}", hex::encode(mac.finalize().into_bytes()));
        if expected != signature {
            return Err(anyhow!("Webhook signature is not valid"));
        }

        Ok(serde_json::from_str(raw_body)?)
    }

    fn request(&self, method: Method, path: &str) -> reqwest::RequestBuilder {
        self.http
            .request(method, format!("{}{}", self.base_url, path))
            .bearer_auth(&self.api_token)
            .header("Accept", "application/json")
    }

    async fn send(&self, request: reqwest::RequestBuilder) -> Result<Value> {
        let response = request.send().await?;
        let status = response.status();
        let payload: Value = response.json().await?;

        if !status.is_success() || !payload["success"].as_bool().unwrap_or(false) {
            return Err(anyhow!(
                "PayGate request failed: {}",
                payload["error"]["message"].as_str().unwrap_or("Unknown error")
            ));
        }

        Ok(payload["data"].clone())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let client = PayGateClient::new(
        "https://paygate.digixsolution.net",
        std::env::var("PAYGATE_STORE_API_TOKEN")?,
    )?;

    let transaction = client
        .charge(
            json!({
                "order_id": "INV-2026-0001",
                "amount": 150000,
                "currency": "IDR",
                "payment_type": "bank_transfer",
                "bank": "bsi",
                "customer": {
                    "name": "Budi",
                    "email": "budi@example.com",
                    "phone": "+628123456789"
                },
                "items": [{
                    "id": "SKU-1",
                    "name": "Kaos PayGate",
                    "qty": 1,
                    "price": 150000
                }]
            }),
            Some("idem_INV-2026-0001"),
        )
        .await?;

    println!("{transaction:#}");
    Ok(())
}
